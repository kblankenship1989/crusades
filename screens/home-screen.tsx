import React from 'react';
import {
    View,
    Text,
    FlatList,
    ListRenderItem,
    TouchableOpacity
} from 'react-native';
import {Icon, Input} from 'react-native-elements';
import {StackNavigationProp} from '@react-navigation/stack';
import {RootParamList} from '../types/root-param-list';
import {OrderOfBattle} from '../types/order-of-battle';
import {v4} from 'uuid';


type HomeState = {
    title: string,
    ordersOfBattle: OrderOfBattle[]
}

type HomeProps = {
    navigation: StackNavigationProp<RootParamList, 'Home'>
}

export class HomeScreen extends React.Component<HomeProps, HomeState> {
    constructor(props : HomeProps) {
        super(props);

        this.state = {
            title: '',
            ordersOfBattle: []
        };
    }

    _orderOfBattleRenderItem : ListRenderItem<OrderOfBattle> = ({item}) : JSX.Element => (
        <TouchableOpacity
            onPress={() => this.props.navigation.navigate('OrderOfBattleSummary', {orderOfBattle: item})}
        >
            <Text>{item.title}</Text>
        </TouchableOpacity>
    )

    addOrderOfBattle = () :void => {
        const title = this.state.title;
        const orderOfBattle : OrderOfBattle = {
            id: v4(),
            title: title,
            faction: ''
        };

        this.setState({
            ordersOfBattle: this.state.ordersOfBattle.concat(orderOfBattle),
            title: ''
        });

        this.props.navigation.navigate('OrderOfBattleSummary', {orderOfBattle});
    }

    render() : JSX.Element {
        return (
            <View>
                <View>
                    <Input
                        leftIcon={
                            <Icon
                                size={18}
                                name={'format-title'}
                                type={'material-community'}
                            />
                        }
                        placeholder={'Title'}
                        onChangeText={(title) => this.setState({title})}
                        value={this.state.title}
                        style={{
                            fontSize: 18,
                            margin: 10
                        }}
                    />
                    <TouchableOpacity
                        onPress={this.addOrderOfBattle}
                        disabled={this.state.title === ''}
                        testID={'create-button'}
                        style={{
                            flex: 1,
                            flexDirection: 'row',
                            backgroundColor: '#404040',
                            alignItems: 'center',
                            justifyContent: 'center',
                            margin: 30
                        }}
                    >
                        <Icon
                            name={'plus'}
                            type={'font-awesome'}
                            size={18}
                            color={'#8ba4c9'}
                            style={{
                                margin: 10
                            }}
                        />
                        <Text style={{
                            color: '#8ba4c9',
                            fontSize: 18,
                            marginVertical: 10
                        }}>{'Create'}</Text>
                    </TouchableOpacity>
                </View>
                <FlatList
                    renderItem={this._orderOfBattleRenderItem}
                    keyExtractor={(item : OrderOfBattle) => item.id}
                    data={this.state.ordersOfBattle}
                />
            </View>
        );
    }
}
