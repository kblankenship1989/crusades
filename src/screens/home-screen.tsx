import React from 'react';
import {FlatList, ListRenderItem} from 'react-native';
import {Text} from 'react-native-elements';
import {ScrollView} from 'react-native-gesture-handler';
import {ConnectedProps} from 'react-redux';
import {StackNavigationProp} from '@react-navigation/stack';

import {homeScreenConnector} from './home-screen-connector';
import {ActionFixedFooterContainer} from '../containers/action-fixed-footer-container';
import {BackgroundImageListItem} from '../components/background-image-list-item';

export type HomeProps = ConnectedProps<typeof homeScreenConnector> & {
    navigation: StackNavigationProp<RootParamList, 'Home'>
}

export const HomeScreen : React.FC<HomeProps> = ({
    loadSelectedOrderOfBattle,
    deleteSelectedOrderOfBattle,
    createOrderOfBattle,
    navigation,
    ordersOfBattle
}) => {
    const selectOrderOfBattle = (index: number) : void => {
        loadSelectedOrderOfBattle(index);
        navigation.push('OrderOfBattleSummary');
    };

    const addOrderOfBattle = () : void => {
        createOrderOfBattle();

        navigation.push('OrderOfBattleSummary');
    };

    const renderListItem : ListRenderItem<OrderOfBattle> = ({item, index}) => (
        <BackgroundImageListItem
            index={index}
            title={item.title || 'Untitled'}
            onPress={selectOrderOfBattle}
            onDelete={deleteSelectedOrderOfBattle}
            imageKey={item.faction}
        />
    );

    return (
        <ActionFixedFooterContainer
            onAdd={addOrderOfBattle}
        >
            {ordersOfBattle.length ?
                <FlatList
                    renderItem={renderListItem}
                    keyExtractor={(orderOfBattle) => orderOfBattle.id.toString()}
                    data={ordersOfBattle}
                />
                :
                <ScrollView>
                    <Text>{'Click + ADD to create a new Crusade Force'}</Text>
                </ScrollView>
            }
        </ActionFixedFooterContainer>
    );
};
