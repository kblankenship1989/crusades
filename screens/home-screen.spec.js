import React from 'react';
import {render, fireEvent} from "@testing-library/react-native";
import {HomeScreen} from "./home-screen";
import {v4} from 'uuid';

jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');
jest.mock('uuid');

const testId = '123';
v4.mockReturnValue(testId);

describe('Given the Home Screen', () => {
    const renderComponent = (overrides) => {
        const requiredProps = {
            navigation: {
                navigate: jest.fn()
            }
        };

        const optionalProps = {};

        const testProps = {
            ...requiredProps,
            ...optionalProps,
            ...overrides
        };

        const component = render(<HomeScreen {...testProps} />);

        return {component, testProps};
    };

    it('should render correctly', () => {
        const {component} = renderComponent();

        expect(component.toJSON()).toMatchSnapshot();
    });

    it('should disable the create button when the title field is blank', () => {
        const {component, testProps} = renderComponent();

        const titleField = component.getByPlaceholderText('Title');
        expect(titleField.props.value).toStrictEqual('');

        const createButton = component.getByTestId('create-button');
        fireEvent(createButton, 'onPress');

        expect(testProps.navigation.navigate).not.toHaveBeenCalled();
    });

    it('should enable the create button once a value is entered into the title field', () => {
        const {component, testProps} = renderComponent();

        const title = 'test text';

        const titleField = component.getByPlaceholderText('Title');
        expect(titleField.props.value).toStrictEqual('');

        fireEvent(titleField, 'onChangeText', title);

        const createButton = component.getByTestId('create-button');
        fireEvent(createButton, 'onPress');

        const expectedOrderOfBattle = {
            id: testId,
            title,
            faction: ''
        };

        expect(testProps.navigation.navigate).toHaveBeenCalledWith('OrderOfBattleSummary', {
            orderOfBattle: expectedOrderOfBattle
        });
    });

    it('should clear the title field afte clicking the create button', () => {
        const {component} = renderComponent();

        const title = 'test text';

        const titleField = component.getByPlaceholderText('Title');
        expect(titleField.props.value).toStrictEqual('');

        fireEvent(titleField, 'onChangeText', title);
        expect(titleField.props.value).toStrictEqual(title);

        const createButton = component.getByTestId('create-button');
        fireEvent(createButton, 'onPress');

        expect(titleField.props.value).toStrictEqual('');
    });
});
