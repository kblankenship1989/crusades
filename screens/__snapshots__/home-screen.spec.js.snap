// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the Home Screen should render correctly 1`] = `
<View>
  <View>
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "overflow": "hidden",
              }
            }
          >
            <View
              theme={
                Object {
                  "colors": Object {
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": Object {
                      "android": Object {
                        "error": "#f44336",
                        "primary": "#2196f3",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "ios": Object {
                        "error": "#ff3b30",
                        "primary": "#007aff",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#8F0CE8",
                    "success": "#52c41a",
                    "warning": "#faad14",
                  },
                }
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
              >
                <Text />
              </View>
            </View>
          </View>
        </View>
        <TextInput
          allowFontScaling={true}
          editable={true}
          onChangeText={[Function]}
          placeholder="Title"
          rejectResponderTermination={true}
          style={
            Object {
              "alignSelf": "center",
              "color": "black",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "margin": 5,
          }
        }
      />
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#404040",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "margin": 30,
          "opacity": 1,
        }
      }
      testID="create-button"
    >
      <View
        style={
          Object {
            "overflow": "hidden",
          }
        }
      >
        <View
          style={
            Object {
              "margin": 10,
            }
          }
          theme={
            Object {
              "colors": Object {
                "disabled": "hsl(208, 8%, 90%)",
                "divider": "#bcbbc1",
                "error": "#ff190c",
                "grey0": "#393e42",
                "grey1": "#43484d",
                "grey2": "#5e6977",
                "grey3": "#86939e",
                "grey4": "#bdc6cf",
                "grey5": "#e1e8ee",
                "greyOutline": "#bbb",
                "platform": Object {
                  "android": Object {
                    "error": "#f44336",
                    "primary": "#2196f3",
                    "secondary": "#9C27B0",
                    "success": "#4caf50",
                    "warning": "#ffeb3b",
                  },
                  "ios": Object {
                    "error": "#ff3b30",
                    "primary": "#007aff",
                    "secondary": "#5856d6",
                    "success": "#4cd964",
                    "warning": "#ffcc00",
                  },
                },
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#8F0CE8",
                "success": "#52c41a",
                "warning": "#faad14",
              },
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
          >
            <Text />
          </View>
        </View>
      </View>
      <Text
        style={
          Object {
            "color": "#8ba4c9",
            "fontSize": 18,
            "marginVertical": 10,
          }
        }
      >
        Create
      </Text>
    </View>
  </View>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</View>
`;
